presto-main/src/main/java/com/facebook/presto/execution/SqlQueryQueueManager.java

```diff
@@ -86,6 +93,61 @@ public SqlQueryQueueManager(QueryManagerConfig config, ObjectMapper mapper, MBea
             }
         }
         this.rules = rules.build();
+        checkIsDAG(this.rules);
+    }
+
+    private void checkIsDAG(List<QueryQueueRule> rules)
+    {
+        DirectedPseudograph<String, DefaultEdge> graph = new DirectedPseudograph<String, DefaultEdge>(DefaultEdge.class);
+        for (QueryQueueRule rule : rules) {
+            String lastQueueName = null;
+            for (QueryQueueDefinition queue : rule.getQueues()) {
+                String currentQueueName = queue.getTemplate();
+                graph.addVertex(currentQueueName);
+                if (lastQueueName != null) {
+                    graph.addEdge(lastQueueName, currentQueueName);
+                }
+                lastQueueName = currentQueueName;
+            }
+        }
+
+        List<String> shortestCycle = shortestCycle(graph);
+
+        if (shortestCycle != null) {
+            String s = Joiner.on(", ").join(shortestCycle);
+            throw new IllegalArgumentException(String.format("Queues must not contain a cycle. The shortest cycle found is [%s]", s));
+        }
+    }
+
+    private static List<String> shortestCycle(DirectedGraph<String, DefaultEdge> g)
+    {
+        FloydWarshallShortestPaths<String, DefaultEdge> floyd = new FloydWarshallShortestPaths<>(g);
+        int minD = Integer.MAX_VALUE;
```

Comment:
We try to avoid abbreviations, so name these `minDistance`, `minSource`, and `minDestination`